File: src/lib/import-export.ts
import { format } from 'date-fns'
import Papa from 'papaparse'
import * as XLSX from 'xlsx'
import { parseDate, toISODate } from '@/lib/dates'
import { parseMoneyInput } from '@/lib/currency'
import type { Category, ImportRowError, MappedImportRow, Transaction, TransactionType } from '@/types'

export const IMPORT_FIELDS = [
  { key: 'date', label: 'Date', required: true },
  { key: 'amount', label: 'Amount', required: true },
  { key: 'type', label: 'Type', required: false },
  { key: 'merchant', label: 'Merchant', required: false },
  { key: 'category', label: 'Category', required: false },
  { key: 'account', label: 'Account', required: false },
  { key: 'description', label: 'Description', required: false },
  { key: 'notes', label: 'Notes', required: false },
  { key: 'paymentMethod', label: 'Payment method', required: false },
  { key: 'tags', label: 'Tags', required: false },
] as const

export type ImportFieldKey = (typeof IMPORT_FIELDS)[number]['key']

export function detectColumns(headers: string[]): Record<ImportFieldKey, string> {
  const mapping = {} as Record<ImportFieldKey, string>
  const normalized = headers.map((header) => header.trim())
  const rules: Array<[ImportFieldKey, string[]]> = [
    ['date', ['date', 'transaction date', 'posted', 'value date']],
    ['amount', ['amount', 'value', 'debit', 'credit', 'sum']],
    ['type', ['type', 'transaction type', 'in/out']],
    ['merchant', ['merchant', 'payee', 'name', 'description']],
    ['category', ['category', 'label']],
    ['account', ['account', 'wallet', 'account name']],
    ['description', ['details', 'memo', 'narration']],
    ['notes', ['notes', 'note', 'comment']],
    ['paymentMethod', ['payment method', 'method', 'mode']],
    ['tags', ['tags', 'labels']],
  ]
  for (const [field, aliases] of rules) {
    const match = normalized.find((header) => aliases.includes(header.toLowerCase()))
    mapping[field] = match ?? ''
  }
  return mapping
}

export function parseTableFile(file: File): Promise<{ headers: string[]; rows: Record<string, string>[] }> {
  const name = file.name.toLowerCase()
  if (name.endsWith('.json')) {
    return file.text().then((text) => {
      const parsed = JSON.parse(text) as unknown
      const rows = Array.isArray(parsed)
        ? parsed
        : ((parsed as { transactions?: unknown[] }).transactions ?? [])
      if (!Array.isArray(rows) || rows.length === 0) {
        throw new Error('JSON must contain an array of transactions.')
      }
      const objects = rows.map((row) => row as Record<string, unknown>)
      const headers = Array.from(new Set(objects.flatMap((row) => Object.keys(row))))
      return {
        headers,
        rows: objects.map((row) =>
          Object.fromEntries(headers.map((header) => [header, String(row[header] ?? '')])),
        ),
      }
    })
  }

  if (name.endsWith('.xlsx') || name.endsWith('.xls')) {
    return file.arrayBuffer().then((buffer) => {
      const workbook = XLSX.read(buffer, { type: 'array' })
      const sheet = workbook.Sheets[workbook.SheetNames[0]]
      const json = XLSX.utils.sheet_to_json<Record<string, unknown>>(sheet, { defval: '' })
      const headers = json.length ? Object.keys(json[0]) : []
      return {
        headers,
        rows: json.map((row) => Object.fromEntries(headers.map((header) => [header, String(row[header] ?? '')]))),
      }
    })
  }

  return new Promise((resolve, reject) => {
    Papa.parse<Record<string, string>>(file, {
      header: true,
      skipEmptyLines: true,
      complete: (result) => {
        if (result.errors.length && !result.data.length) {
          reject(new Error(result.errors[0].message))
          return
        }
        resolve({
          headers: result.meta.fields ?? [],
          rows: result.data,
        })
      },
      error: (error) => reject(error),
    })
  })
}

function inferType(raw: string, amount: number): TransactionType {
  const value = raw.toLowerCase()
  if (value.includes('income') || value.includes('credit') || amount < 0 && value.includes('in')) return 'income'
  if (value.includes('transfer')) return 'transfer'
  if (amount < 0) return 'expense'
  if (value.includes('expense') || value.includes('debit')) return 'expense'
  return amount >= 0 ? 'expense' : 'income'
}

export function mapImportRows(
  rows: Record<string, string>[],
  mapping: Record<ImportFieldKey, string>,
  options: {
    defaultAccountId: string
    defaultCurrency: string
    categories: Category[]
    accountNames: Record<string, string>
  },
): MappedImportRow[] {
  return rows.map((row, index) => {
    const errors: ImportRowError[] = []
    const dateRaw = mapping.date ? row[mapping.date] : ''
    const amountRaw = mapping.amount ? row[mapping.amount] : ''
    const date = parseDate(dateRaw)
    const amount = parseMoneyInput(amountRaw)
    if (!date) errors.push({ row: index + 1, field: 'date', message: 'Invalid date' })
    if (amount === null) errors.push({ row: index + 1, field: 'amount', message: 'Invalid amount' })

    const typeRaw = mapping.type ? row[mapping.type] : ''
    const type = inferType(typeRaw, amount ?? 0)
    const categoryName = mapping.category ? row[mapping.category] : ''
    const category = options.categories.find(
      (item) => item.name.toLowerCase() === categoryName.trim().toLowerCase(),
    )
    const accountName = mapping.account ? row[mapping.account] : ''
    const accountId =
      Object.entries(options.accountNames).find(([, name]) => name.toLowerCase() === accountName.trim().toLowerCase())?.[0] ??
      options.defaultAccountId

    const paymentRaw = (mapping.paymentMethod ? row[mapping.paymentMethod] : 'card').toLowerCase()
    const paymentMethod = (
      ['cash', 'card', 'upi', 'bank_transfer', 'wallet', 'other'] as const
    ).includes(paymentRaw as never)
      ? (paymentRaw as Transaction['paymentMethod'])
      : 'other'

    return {
      row: index + 1,
      errors,
      transaction: {
        type,
        amount: Math.abs(amount ?? 0),
        currency: options.defaultCurrency,
        categoryId: category?.id ?? null,
        subcategoryId: null,
        accountId,
        toAccountId: null,
        merchant: mapping.merchant ? row[mapping.merchant] : '',
        description: mapping.description ? row[mapping.description] : '',
        notes: mapping.notes ? row[mapping.notes] : '',
        date: date ? toISODate(date) : '',
        paymentMethod,
        tags: mapping.tags
          ? row[mapping.tags]
              .split(',')
              .map((tag) => tag.trim())
              .filter(Boolean)
          : [],
        recurringId: null,
        isSample: false,
      },
    }
  })
}

export function transactionsToCsv(transactions: Transaction[]): string {
  return Papa.unparse(
    transactions.map((tx) => ({
      date: tx.date,
      type: tx.type,
      amount: tx.amount,
      currency: tx.currency,
      merchant: tx.merchant,
      description: tx.description,
      notes: tx.notes,
      payment_method: tx.paymentMethod,
      tags: tx.tags.join(','),
    })),
  )
}

export function transactionsToXlsx(transactions: Transaction[]): Blob {
  const sheet = XLSX.utils.json_to_sheet(
    transactions.map((tx) => ({
      Date: tx.date,
      Type: tx.type,
      Amount: tx.amount,
      Currency: tx.currency,
      Merchant: tx.merchant,
      Description: tx.description,
      Notes: tx.notes,
      PaymentMethod: tx.paymentMethod,
      Tags: tx.tags.join(','),
    })),
  )
  const workbook = XLSX.utils.book_new()
  XLSX.utils.book_append_sheet(workbook, sheet, 'Transactions')
  const buffer = XLSX.write(workbook, { type: 'array', bookType: 'xlsx' })
  return new Blob([buffer], {
    type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
  })
}

export function exportFilename(kind: string, extension: string): string {
  return `aureum-${kind}-${format(new Date(), 'yyyy-MM-dd')}.${extension}`
}

File: flutter/lib/core/import_export.dart
import 'package:csv/csv.dart';
import 'package:intl/intl.dart';

import '../models/models.dart';
import 'currency.dart';
import 'dates.dart';

class ImportField {
  const ImportField({required this.key, required this.label, required this.requiredField});

  final String key;
  final String label;
  final bool requiredField;
}

const importFields = [
  ImportField(key: 'date', label: 'Date', requiredField: true),
  ImportField(key: 'amount', label: 'Amount', requiredField: true),
  ImportField(key: 'type', label: 'Type', requiredField: false),
  ImportField(key: 'merchant', label: 'Merchant', requiredField: false),
  ImportField(key: 'category', label: 'Category', requiredField: false),
  ImportField(key: 'account', label: 'Account', requiredField: false),
  ImportField(key: 'description', label: 'Description', requiredField: false),
  ImportField(key: 'notes', label: 'Notes', requiredField: false),
  ImportField(key: 'paymentMethod', label: 'Payment method', requiredField: false),
  ImportField(key: 'tags', label: 'Tags', requiredField: false),
];

Map<String, String> detectColumns(List<String> headers) {
  final mapping = <String, String>{for (final field in importFields) field.key: ''};
  final normalized = headers.map((header) => header.trim()).toList();
  const rules = <String, List<String>>{
    'date': ['date', 'transaction date', 'posted', 'value date'],
    'amount': ['amount', 'value', 'debit', 'credit', 'sum'],
    'type': ['type', 'transaction type', 'in/out'],
    'merchant': ['merchant', 'payee', 'name', 'description'],
    'category': ['category', 'label'],
    'account': ['account', 'wallet', 'account name'],
    'description': ['details', 'memo', 'narration'],
    'notes': ['notes', 'note', 'comment'],
    'paymentMethod': ['payment method', 'method', 'mode'],
    'tags': ['tags', 'labels'],
  };
  for (final entry in rules.entries) {
    final match = normalized.cast<String?>().firstWhere(
          (header) => entry.value.contains(header!.toLowerCase()),
          orElse: () => null,
        );
    mapping[entry.key] = match ?? '';
  }
  return mapping;
}

({List<String> headers, List<Map<String, String>> rows}) parseCsvTable(String text) {
  final table = Csv().decode(text);
  if (table.isEmpty) return (headers: <String>[], rows: <Map<String, String>>[]);
  final headers = table.first.map((cell) => '$cell').toList();
  final rows = table.skip(1).map((row) {
    final map = <String, String>{};
    for (var i = 0; i < headers.length; i++) {
      map[headers[i]] = i < row.length ? '${row[i]}' : '';
    }
    return map;
  }).toList();
  return (headers: headers, rows: rows);
}

TransactionType inferType(String raw, double amount) {
  final value = raw.toLowerCase();
  if (value.contains('income') || value.contains('credit') || (amount < 0 && value.contains('in'))) {
    return TransactionType.income;
  }
  if (value.contains('transfer')) return TransactionType.transfer;
  if (amount < 0) return TransactionType.expense;
  if (value.contains('expense') || value.contains('debit')) return TransactionType.expense;
  return amount >= 0 ? TransactionType.expense : TransactionType.income;
}

List<MappedImportRow> mapImportRows(
  List<Map<String, String>> rows,
  Map<String, String> mapping, {
  required String defaultAccountId,
  required String defaultCurrency,
  required List<Category> categories,
  required Map<String, String> accountNames,
}) {
  return [
    for (var index = 0; index < rows.length; index++)
      () {
        final row = rows[index];
        final errors = <ImportRowError>[];
        final dateRaw = mapping['date']?.isNotEmpty == true ? row[mapping['date']] ?? '' : '';
        final amountRaw = mapping['amount']?.isNotEmpty == true ? row[mapping['amount']] ?? '' : '';
        final date = parseDate(dateRaw);
        final amount = parseMoneyInput(amountRaw);
        if (date == null) errors.add(ImportRowError(row: index + 1, field: 'date', message: 'Invalid date'));
        if (amount == null) errors.add(ImportRowError(row: index + 1, field: 'amount', message: 'Invalid amount'));

        final typeRaw = mapping['type']?.isNotEmpty == true ? row[mapping['type']] ?? '' : '';
        final type = inferType(typeRaw, amount ?? 0);
        final categoryName = mapping['category']?.isNotEmpty == true ? row[mapping['category']] ?? '' : '';
        Category? category;
        for (final item in categories) {
          if (item.name.toLowerCase() == categoryName.trim().toLowerCase()) {
            category = item;
            break;
          }
        }
        final accountName = mapping['account']?.isNotEmpty == true ? row[mapping['account']] ?? '' : '';
        String accountId = defaultAccountId;
        accountNames.forEach((id, name) {
          if (name.toLowerCase() == accountName.trim().toLowerCase()) accountId = id;
        });

        final paymentRaw =
            (mapping['paymentMethod']?.isNotEmpty == true ? row[mapping['paymentMethod']] ?? 'card' : 'card')
                .toLowerCase();
        final paymentMethod = PaymentMethodX.parse(paymentRaw);

        return MappedImportRow(
          row: index + 1,
          errors: errors,
          transaction: CreateTransactionInput(
            type: type,
            amount: (amount ?? 0).abs(),
            currency: defaultCurrency,
            categoryId: category?.id,
            subcategoryId: null,
            accountId: accountId,
            toAccountId: null,
            merchant: mapping['merchant']?.isNotEmpty == true ? row[mapping['merchant']] ?? '' : '',
            description: mapping['description']?.isNotEmpty == true ? row[mapping['description']] ?? '' : '',
            notes: mapping['notes']?.isNotEmpty == true ? row[mapping['notes']] ?? '' : '',
            date: date == null ? '' : toISODate(date),
            paymentMethod: paymentMethod,
            tags: mapping['tags']?.isNotEmpty == true
                ? (row[mapping['tags']] ?? '').split(',').map((tag) => tag.trim()).where((tag) => tag.isNotEmpty).toList()
                : const [],
            recurringId: null,
            attachmentPath: null,
            attachmentName: null,
            isSample: false,
          ),
        );
      }(),
  ];
}

String transactionsToCsv(List<Transaction> transactions) {
  final rows = [
    ['date', 'type', 'amount', 'currency', 'merchant', 'description', 'notes', 'payment_method', 'tags'],
    ...transactions.map(
      (tx) => [
        tx.date,
        tx.type.value,
        tx.amount,
        tx.currency,
        tx.merchant,
        tx.description,
        tx.notes,
        tx.paymentMethod.value,
        tx.tags.join(','),
      ],
    ),
  ];
  return Csv().encode(rows);
}

String exportFilename(String kind, String extension) {
  return 'aureum-$kind-${DateFormat('yyyy-MM-dd').format(DateTime.now())}.$extension';
}
